diagnostics/reporter.rs

```diff
@@ -0,0 +1,154 @@
+    deployment_id: String,
+    diagnostics: Arc<Diagnostics>,
+    reporting_uri: &'static str,
+    data_directory: PathBuf,
+    is_enabled: bool,
+    _reporting_job: Arc<Option<IntervalRunner>>,
+}
+
+impl Reporter {
+    pub(crate) fn new(
+        deployment_id: String,
+        diagnostics: Arc<Diagnostics>,
+        reporting_uri: &'static str,
+        data_directory: PathBuf,
+        is_enabled: bool,
+    ) -> Self {
+        Self { deployment_id, diagnostics, reporting_uri, data_directory, is_enabled, _reporting_job: Arc::new(None) }
+    }
+
+    pub fn may_start(&self) {
+        if self.is_enabled {
+            Self::delete_disabled_reporting_file_if_exists(&self.data_directory);
+            self.schedule_reporting();
+        } else {
+            self.report_once_if_needed();
+        }
+    }
+
+    fn schedule_reporting(&self) {
+        let diagnostics = self.diagnostics.clone();
+        let reporting_uri = self.reporting_uri;
+
+        let reporting_job = IntervalRunner::new_with_initial_delay(
+            move || {
+                Self::report(diagnostics.clone(), reporting_uri);
+            },
+            REPORT_INTERVAL,
+            self.calculate_initial_delay(),
+        );
+        let _ = Arc::get_mut(&mut Arc::clone(&self._reporting_job)).map(|opt| *opt = Some(reporting_job));
+    }
+
+    fn report_once_if_needed(&self) {
+        let disabled_reporting_file = self.data_directory.join(DISABLED_REPORTING_FILE_NAME);
+        if !disabled_reporting_file.exists() {
+            let diagnostics = self.diagnostics.clone();
+            let reporting_uri = self.reporting_uri;
+            let data_directory = self.data_directory.clone();
+            thread::spawn(move || {
+                thread::sleep(Duration::from_secs(3600)); // TODO: Make a constant
+                if Self::report(diagnostics, reporting_uri) {
+                    Self::save_disabled_reporting_file(&data_directory);
+                }
+            });
+        }
+    }
+
+    fn calculate_initial_delay(&self) -> Duration {
+        let report_interval_secs = REPORT_INTERVAL.as_secs();
+        assert!(report_interval_secs == 3600, "Modify the algorithm if you change the interval!");
+
+        let current_minute = Local::now().minute() as u64;
+
+        let mut deployment_id_hasher = DefaultHasher::new();
+        self.deployment_id.hash(&mut deployment_id_hasher);
+        let scheduled_minute = deployment_id_hasher.finish() % report_interval_secs / 60;
+
+        let delay_secs = if current_minute > scheduled_minute {
+            report_interval_secs - (current_minute + scheduled_minute) * 60
+        } else {
+            (scheduled_minute - current_minute) * 60
+        };
+
+        Duration::from_secs(delay_secs)
+    }
+
+    fn report(diagnostics: Arc<Diagnostics>, reporting_uri: &'static str) -> bool {
+        let diagnostics_json = diagnostics.to_reporting_json().to_string();
+        diagnostics.take_snapshot();
```

Comment:
Are these the wrong way around? Shouldn't we take a snapshot and report that snapshot?

Reply:
I think I'm still confused as to what `take_snapshot` is actually doing. It could use a better name.

Is it more like `reset` if the deltas are calculated against the snapshot, so effectively the deltas are reset to 0?

Reply:
Uhhhmmmmmm `reset` sounds scarier and makes it look like we forget the whole data! I'll start by renaming `diagnsotics.to_reporting_json` to `to_reporting_json_against_snapshot`. It should give a better hint of the workflow and what's expected inside.